File: src/bindings/kotlin/templates/ObjectTemplate.kt
class {{ obj.name()|class_name_kt }}(handle: Long) {
    private var handle: AtomicLong = AtomicLong(handle)
    {%- for cons in obj.constructors() %}
    constructor({% for arg in cons.arguments() %}{{ arg.name() }}: {{ arg.type_()|type_kt }}{% if loop.last %}{% else %}, {% endif %}{% endfor %}) :
        this(
            _UniFFILib.INSTANCE.{{ cons.ffi_func().name() }}(
                {%- for arg in cons.arguments() %}
                {{ arg.name()|lower_kt(arg.type_()) }}{% if loop.last %}{% else %},{% endif %}
                {%- endfor %}
            )
        )
    {%- endfor %}

    // XXX TODO: destructors or equivalent.

    {%- for meth in obj.methods() %}
    fun {{ meth.name()|fn_name_kt }}(
        {% for arg in meth.arguments() %}
        {{ arg.name() }}: {{ arg.type_()|type_kt }}{% if loop.last %}{% else %}, {% endif %}
        {% endfor %}
    ): {% match meth.return_type() %}{% when Some with (type_) %}{{ type_|type_kt }}{% when None %}Unit{% endmatch %} {
        val _retval = _UniFFILib.INSTANCE.{{ meth.ffi_func().name() }}(
            this.handle.get(){% if meth.arguments().len() > 0 %},{% endif %}
            {%- for arg in meth.arguments() %}
            {{ arg.name()|lower_kt(arg.type_()) }}{% if loop.last %}{% else %},{% endif %}
            {%- endfor %}
        )
        {% match meth.return_type() %}{% when Some with (return_type) %}return {{ "_retval"|lift_kt(return_type) }}{% else %}return _retval{% endmatch %}
    }
    {%- endfor %}
}

File: uniffi_bindgen/src/bindings/rust/templates/ObjectTemplate.rs
pub struct {{ obj.name()|class_name_rs }};

impl {{ obj.name()|class_name_rs }} {
  {%- match obj.primary_constructor() %}
  {%- when Some with (cons) %}
  pub fn new({% call rs::arg_list_decl(cons.arguments()) -%}) -> {% call rs::fallible_return_type(cons, "Self") %} {
    todo!()
  }

  {%- when None %}
  {%- endmatch %}

  {% for cons in obj.alternate_constructors() -%}
  pub fn {{ cons.name()|fn_name_rs }}({% call rs::arg_list_decl(cons.arguments()) %}) -> {% call rs::fallible_return_type(cons, "Self") %} {
    todo!()
  }
  {% endfor %}

  {% for meth in obj.methods() -%}
  {%- match meth.return_type() -%}

  {%- when Some with (return_type) -%}
  pub fn {{ meth.name()|fn_name_rs }}({% call rs::self_arg(meth) %}, {% call rs::arg_list_decl(meth.arguments()) %}) -> {% call rs::fallible_return_type(meth, return_type|type_rs) %} {
    todo!()
  }

  {%- when None -%}
  pub fn {{ meth.name()|fn_name_rs }}({% call rs::self_arg(meth) %}, {% call rs::arg_list_decl(meth.arguments()) %}) -> {% call rs::fallible_return_type(meth, "()") %} {
    todo!()
  }
  {% endmatch %}
  {% endfor %}
}
